### PR TITLE
[Java] Enable direct call by default.

### PR DESCRIPTION
**Pull request by jovany-wang**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: java/api/src/main/java/org/ray/api/options/CallOptions.java [hunk 1]
```diff
@@ -18,23 +18,14 @@ private CallOptions(Map<String, Double> resources, boolean useDirectCall) {
...
 
     private Map<String, Double> resources = new HashMap<>();
-    private boolean useDirectCall = DEFAULT_USE_DIRECT_CALL;
 
     public Builder setResources(Map<String, Double> resources) {
       this.resources = resources;
       return this;
     }
 
-    // Since direct call is not fully supported yet (see issue #5559),
```
Comment (kfstorm) [inline java/api/src/main/java/org/ray/api/options/CallOptions.java]: There's another copy of these comments in `ActorCreationOptions.java`. Please remove it too.

File: java/test/src/main/java/org/ray/api/test/ActorReconstructionTest.java [hunk 2]
```diff
@@ -63,9 +63,9 @@ public void testActorReconstruction() throws InterruptedException, IOException {
     // Wait for the actor to be killed.
     TimeUnit.SECONDS.sleep(1);
 
-    // Try calling increase on this actor again and check the value is now 4.
+    // TODO(qwang): Check
```
Comment (kfstorm) [inline java/test/src/main/java/org/ray/api/test/ActorReconstructionTest.java]: What's this TODO for?

File: java/test/src/main/java/org/ray/api/test/ActorTest.java [hunk 3]
```diff
@@ -129,9 +132,12 @@ public void testPassActorAsParameter() {
 
   public void testUnreconstructableActorObject() throws InterruptedException {
     TestUtils.skipTestUnderSingleProcess();
+    if (((AbstractRayRuntime) Ray.internal()).getRayConfig().runMode != RunMode.SINGLE_PROCESS) {
+      throw new SkipException("1111111111111111111");
```
Comment (kfstorm) [inline java/test/src/main/java/org/ray/api/test/ActorTest.java]: Give a reasonable message.

File: java/test/src/main/java/org/ray/api/test/ActorTest.java [hunk 4]
```diff
@@ -129,9 +132,12 @@ public void testPassActorAsParameter() {
...
+    }
+
     // The UnreconstructableException is created by raylet.
-    // TODO (kfstorm): This should be supported by direct actor call.
-    TestUtils.skipTestIfDirectActorCallEnabled();
+    // TODO(qwang):
```
Comment (kfstorm) [inline java/test/src/main/java/org/ray/api/test/ActorTest.java]:
```suggestion
    // TODO(qwang): Will re-enable this test case once ref counting is supported in Java.
```

File: java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java [hunk 5]
```diff
@@ -63,7 +63,6 @@ public void testPythonCallJavaFunction() {
   @Test
   public void testCallingPythonActor() {
     // Python worker doesn't support direct call yet.
```
Comment (kfstorm) [inline java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java]: Remove this comment.

File: java/test/src/main/java/org/ray/api/test/FailureTest.java [hunk 6]
```diff
@@ -121,8 +121,6 @@ public void testWorkerProcessDying() {
   public void testActorProcessDying() {
     TestUtils.skipTestUnderSingleProcess();
     // This test case hangs if the worker to worker connection is implemented with grpc.
```
Comment (kfstorm) [inline java/test/src/main/java/org/ray/api/test/FailureTest.java]: Remove this comment.

**Review by kfstorm — COMMENTED** [review]
Please search for `isDirectCall` or `useDirectCall` in the codebase. I believe there are some places you missed, like `ArgumentsBuilder.wrap`.

File: java/runtime/src/main/java/org/ray/runtime/RayNativeRuntime.java [hunk 7]
```diff
@@ -137,10 +136,7 @@ public void setResource(String resourceName, double capacity, UniqueId nodeId) {
 
   @Override
   public void killActor(RayActor<?> actor) {
-    if (!((NativeRayActor) actor).isDirectCallActor()) {
-      throw new UnsupportedOperationException("Only direct call actors can be killed.");
-    }
-    nativeKillActor(nativeCoreWorkerPointer, actor.getId().getBytes());
+    throw new UnsupportedOperationException("Only direct call actors can be killed.");
```
Comment (kfstorm) [inline java/runtime/src/main/java/org/ray/runtime/RayNativeRuntime.java]:
```suggestion
    nativeKillActor(nativeCoreWorkerPointer, actor.getId().getBytes());
```

File: java/test/src/main/java/org/ray/api/test/PlasmaFreeTest.java [hunk 8]
```diff
@@ -25,11 +26,11 @@ public void testDeleteObjects() {
...
         !TestUtils.getRuntime().getObjectStore()
           .wait(ImmutableList.of(helloId.getId()), 1, 0).get(0), 50);
-    if (TestUtils.isDirectActorCallEnabled()) {
-      // Direct call will not delete object from im-memory store.
-      Assert.assertFalse(result);
-    } else {
+    if (TestUtils.isSingleProcessMode()) {
       Assert.assertTrue(result);
+    } else {
+      // The cluster mode use ref-counting to delete the object.
```
Comment (jovany-wang) [inline java/test/src/main/java/org/ray/api/test/PlasmaFreeTest.java]:
```suggestion
      // The cluster mode uses ref-counting to delete the objects.
```

**Comment by jovany-wang** [comment]
@kfstorm Comments were addressed.

**Review by kfstorm — APPROVED** [review]
LGTM. Thanks!

File: java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java [hunk 9]
```diff
@@ -32,27 +28,17 @@
...
-        if (!isDirectCall && value.data.length > LARGEST_SIZE_PASS_BY_VALUE) {
-          RayRuntime runtime = Ray.internal();
-          if (runtime instanceof RayMultiWorkerNativeRuntime) {
-            runtime = ((RayMultiWorkerNativeRuntime) runtime).getCurrentRuntime();
-          }
-          id = ((AbstractRayRuntime) runtime).getObjectStore()
-              .putRaw(value);
-          value = null;
+        if (value.data.length > LARGEST_SIZE_PASS_BY_VALUE) {
+          // Do nothing since we are not support pass by reference in direct call.
```
Comment (raulchen) [inline java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java]: I think this is already supported in Python?
does this mean that we pass everything by value? I think there might be problems with this.

Comment (jovany-wang) [reply java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java]: Yes, we pass all by value on the current master codebase.

If it has been supported in python, I'd like to post another PR to fix it, but this PR aims to remove direct call concept and do not change other logic any more.

Comment (jovany-wang) [reply java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java]: I created an issue to take a note. https://github.com/ray-project/ray/issues/7586

Comment (raulchen) [reply java/runtime/src/main/java/org/ray/runtime/task/ArgumentsBuilder.java]: ok, but this should be fixed ASAP.

**Pull request merged**